Readme
Notes for ARC 5.21e					October 31, 1991

	ARC 5.21 for Unix is currently at patch level 6. The code provided
here has been used on 4.3 BSD based Unix systems and on the Atari ST. (The
GNU C compiler, version 1.40, was used to make the ST version. It should
still compile with Mark Williams C, but I haven't bothered to try it.) The
current version is over twice as fast as previous releases, for all file
compression and decompression operations. It also fixes longstanding bugs
in the squash code and in the tmclock DST code.
 
	The Arcinfo file describes the ARC 5.xx directory format. The 6.xx
format contains additional information that is not supported here.

	System V support in the shar file Sysvarcstuf was provided by Jon
Zeeff, Janet Walz, and Rich Salz. You will probably also need the dirent
directory library written by Doug Gwyn; this can be obtained from uunet or
many other comp.sources.unix archive sites.

	The tmclock.c file was stolen from Jef Poskanzer's tws library,
as distributed in PHOON, PHase of the mOON, and slightly hacked by me.

	The man page was originally from Leo Wilson, with updates by me.

         /                 Howard Chu
   ___  /_ , ,_.     Jet Propulsion Laboratory
       / /(_/(__     hyc@hanauma.jpl.nasa.gov
           /

Notes for ARC 5.21                                        June 6, 1988
 
This program is based on the MSDOS ARC program, version 5.21, plus
a few enhancements... 
 
 o ARC also performs Huffman Squeezing on data. The Huffman Squeeze
   algorithm was removed from MSDOS ARC after version 5.12. It turns
   out to be more efficient than Lempel-Ziv style compression when
   compressing graphic images. Squeeze analysis is always done now,
   and the best of packing, squeezing, or crunching is used.
 
 o Compresses and extracts Squashed files. "Squashing" was created
   by Phil Katz in his PKxxx series of ARC utility programs for
   MSDOS. Dan Lanciani wrote the original modifications to ARC's
   Crunch code to handle Squashing. I've made minor changes since
   then, mostly to reduce the amount of memory required. The 'q'
   option flag must be specified to Squash files. The Squashing
   algorithm will be used instead of the usual Crunch algorithm,
   and will be compared against packing and squeezing, as before.
 
System specific notes:
 
   On MTS, an additional option flag, 'i' for "image mode," was
used. ARC assumes files are text, by default, and will translate
MTS files from EBCDIC to ASCII before storing in an archive, and
translates from ASCII to EBCDIC upon extraction. Specifying the
'i' flag will inhibit this translation. This would most commonly
be used when shipping binary images such as TeX DVI files, other
.ARC files stored within an archive, etc... The 'r' (run) command
is omitted. It just doesn't seem very useful. Also, ARC cannot
restore MTS files with their original time stamps. (Maybe in a
future release...)
 
   On Unix(tm) systems, the 'i' flag is also present. Unix ARC
assumes a binary file, by default. Here the only translation
involved is in end-of-line processing. When storing text files, ARC will
change '\n' to '\r\n', and does the opposite when extracting files.
Carriage returns in any other location are preserved when extracting.
This translation only occurs if the 'i' flag is given.

   On the Atari ST, the 'h' (for "hold screen") option is present,
which simply delays exiting the program. This is typically used when
executing ARC from the desktop, to allow reading all of ARC's output
before the screen is cleared and the desktop is redrawn. The program
will prompt and wait for a keypress before exiting. Note that since
there are no "options" for the MARC program, the "hold screen" option
is always active for MARC.
 
   On both Unix and Atari systems, ARC & MARC will search for an
environment variable named "ARCTEMP" or "TMPDIR." If present, any
temporary files will be created in the specified directory. This is
probably insignificant for Unix users, but can be handy on the Atari,
in combination with a RAMdisk. Highly recommended for floppy users.
(Unfortunately, you can only take advantage of this when running some
form of command shell that allows setting environment variables. Thus,
you won't see any speed gains when running from the desktop.)
 
That about covers things. The enclosed documentation is taken directly
from the MSDOS distribution of ARC. Unless specified differently here,
the programs behave indentically. Note that ARC521.DOC is identical to
ARC520.DOC - the differences between the two versions are described in
the file CHANGES.521.
 
Oh yeah - this program may be distributed freely under the terms of
the GNU General Public License (See COPYING for license details).

It'd be nice if you kept this and the other enclosed doc files with it when
distributing, but I'm not going to make a fuss about it. Most people are so
familiar with the program by now that it wouldn't matter much anyway. You
should keep this README file around, so bug reports & such will find
their way back to me. (Bugs? What bugs? Nah, there aren't any bugs...)
 
         /                 Howard Chu
   ___  /_ , ,_.     University of Michigan
       / /(_/(__     hyc@umix.cc.umich.edu
           /                umix!hyc
          '     


Changelog
                        CHANGES IN VERSION 5.21q
        (Anibal Monsalve Salazar <anibal@debian.org>, 27 Jun 2013)

Relicense under the terms of the General Public License.  See:
  - http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=641449
  - https://bugzilla.redhat.com/show_bug.cgi?id=947786
Remove old csv header lines

                        CHANGES IN VERSION 5.21p
                 (Klaus Reimer <k@ailis.de>, 07 Aug 2010)

Replaced mktemp with mkstemp.
Set correct version number and updated creation date in command-line help.
Corrected content of PATCHLEVEL file.
Added man page for marc.
Added install make target.
Added -Wall compiler flag.


                        CHANGES IN VERSION 5.21o
                 (Howard Chu <hyc@highlandsun.com>, 12 Oct 2005)

Silence warnings about unused results.


                        CHANGES IN VERSION 5.21n
                 (Howard Chu <hyc@highlandsun.com>, 08 Oct 2005)

Better fix for temporary file creations, previous patch did not fix marc.
Silence all gcc-4.0.0 -Wall warnings.


                        CHANGES IN VERSION 5.21m
                 (Klaus Reimer <k@ailis.de>, 21 Sep 2005)

Applied patch from Joey Schulze fixing the following problems:
  
  - Fixed insecure permissions of a temporary file [arcsvc.c, CAN-2005-2945]
  - Fixed insecure temporary file creation [arcsvc.c, CAN-2005-2992]



                        CHANGES IN VERSION 5.21l
                 (Klaus Reimer <k@ailis.de>, 13 Mar 2005)
                        
Fixed warnings because missing inclusion of string.h and fixed a gcc-4.0
problem with marc.c (Missing declaration of makefnam).



                        CHANGES IN VERSION 5.21k
                 (Klaus Reimer <k@ailis.de>, 6 Nov 2004)
                        
Ulf Harnhammar found some buffer overflows causing segfaults and provided a
patch which is now applied to arc to close these bugs.



                        CHANGES IN VERSION 5.21


The following changes have been made in ARC version 5.21, but have not
yet been included in the manual:

o   When adding files to an archive, ARC now shows the stowage facter
    achieved on each file.

o   A bug was found that would keep an archive entry from being
    encrypted if it was stored without compression.  This has now been
    fixed.

o   If changes are made to a corrupted archive, the corrupted entries
    are discarded.  This makes it possible to lose data accidentally.
    ARC will now only make changes to a corrupted archive if the W
    (suppress Warnings) option has been given.

o   The N (suppress Notes) option now suppresses the "Creating
    archive" note when a new archive is being created.

o   The N (suppress Notes) option formerly did nothing useful when
    used with the L (List files) command.  It now causes a terse
    listing of filenames only, suitable for use with pipes and
    redirection.

o   The list of filenames given to ARC may now include indirect
    references.  If a filename begins with an "at sign" ("@"), it is
    taken to be the name of a file which contains a list of file
    names.  The list of file names may include further indirection.
    If no extension is given, ".CMD" is assumed.  For example, the
    command:

         arc a waste junk.txt @trash

    would cause ARC to add JUNK.TXT plus all files listed in the file
    TRASH.CMD to an archive named WASTE.ARC.  If no file is specified,
    then the list is read from standard input.  For example, the
    command:

         arc ln waste | arc a trash @

    would cause ARC to add files to TRASH.ARC based on the names of
    the files stored in WASTE.ARC.  It is probably a good idea to give
    the O (Overwrite) option if you are extracting files this way.

    Version 5.21 of MARC also allows for indirection.


arc.c
/*  ARC - Archive utility
  
    Version 5.21, created on 04/22/87 at 15:05:21
  
    (C) COPYRIGHT 1985-87 by System Enhancement Associates.
    You may copy and distribute this program freely,
    under the terms of the General Public License.
  
    By:	 Thom Henderson
  
    Description:
	 This program is a general archive utility, and is used to maintain
	 an archive of files.  An "archive" is a single file that combines
	 many files, reducing storage space and allowing multiple files to
	 be handled as one.
  
    Instructions:
	 Run this program with no arguments for complete instructions.
  
    Programming notes:
	 ARC Version 2 differs from version 1 in that archive entries
	 are automatically compressed when they are added to the archive,
	 making a separate compression step unecessary.	 The nature of the
	 compression is indicated by the header version number placed in
	 each archive entry, as follows:
  
	 1 = Old style, no compression
	 2 = New style, no compression
	 3 = Compression of repeated characters only
	 4 = Compression of repeated characters plus Huffman SQueezing
	 5 = Lempel-Zev packing of repeated strings (old style)
	 6 = Lempel-Zev packing of repeated strings (new style)
	 7 = Lempel-Zev Williams packing with improved hash function
	 8 = Dynamic Lempel-Zev packing with adaptive reset
	 9 = Dynamic Lempel-Zev packing, larger hash table
  
	 Type 5, Lempel-Zev packing, was added as of version 4.0
  
	 Type 6 is Lempel-Zev packing where runs of repeated characters
	 have been collapsed, and was added as of version 4.1
  
	 Type 7 is a variation of Lempel-Zev using a different hash
	 function which yields speed improvements of 20-25%, and was
	 added as of version 4.6
  
	 Type 8 is a different implementation of Lempel-Zev, using a
	 variable code size and an adaptive block reset, and was added
	 as of version 5.0
  
	 Type 9 is a slight modification of type 8, first used by Phil
	 Katz in his PKARC utilites. The primary difference is the use
	 of a hash table twice as large as for type 8, and that this
	 algorithm called Squashing, doesn't perform run-length encoding
	 on the input data.
  
	 Verion 4.3 introduced a temporary file for holding the result
	 of the first crunch pass, thus speeding up crunching.
  
	 Version 4.4 introduced the ARCTEMP environment string, so that
	 the temporary crunch file may be placed on a ramdisk.	Also
	 added was the distinction bewteen Adding a file in all cases,
	 and Updating a file only if the disk file is newer than the
	 corresponding archive entry.
  
	 The compression method to use is determined when the file is
	 added, based on whichever method yields the smallest result.
  
    Language:
	 Computer Innovations Optimizing C86
*/
#include <stdio.h>
#include "arc.h"

#if	UNIX
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#endif

#include <string.h>
#if BSD
#include <strings.h>
#endif

#if	!__STDC__
char		*calloc(), *malloc(), *realloc();
#endif

VOID		addarc(), delarc(), extarc(), lstarc(), tstarc(), cvtarc(), runarc();
VOID		arcdie();
static	VOID	expandlst();
#if	_MTS
VOID		etoa();
#endif
#if	GEMDOS
long		_stksize = 30720;
#endif
char		*makefnam();	/* filename fixup routine */

	/* Block I/O buffers */

u_char		*pinbuf, *pakbuf, *outbuf, *outend;

static char   **lst;		/* files list */
static int	lnum;		/* length of files list */

int
main(num, arg)			/* system entry point */
	int		num;	/* number of arguments */
	char	       *arg[];	/* pointers to arguments */
{
	char		opt = 0;/* selected action */
	char	       *a;	/* option pointer */
	VOID		upper();/* case conversion routine */
	char	       *envfind();	/* environment searcher */
	int		n;	/* index */
	char	       *arctemp2;
#if	GEMDOS
	VOID		exitpause();
	int		append;
#endif
#if	_MTS
	fortran VOID	guinfo();
	char		gotinf[4];
#endif
#if	UNIX
	struct	stat	sbuf;
#endif

	if (num < 3) {
		printf("ARC - Archive utility, Version 5.21q, created on 06/27/2013\n");
/*		printf("(C) COPYRIGHT 1985,86,87 by System Enhancement Associates\n");
		printf("Please refer all inquiries to:\n\n");
		printf("       System Enhancement Associates\n");
		printf("       21 New Street, Wayne NJ 07470\n\n");
		printf("You may copy and distribute this program freely,");
		printf(" under the terms of the General Public License.\n");
*/
#if	MSDOS
		printf("Usage: ARC {amufdxerplvtc}[bswnoq][g<password>]");
#endif
#if	GEMDOS
		printf("Usage: ARC {amufdxerplvtc}[bhswnoq][g<password>]");
#endif
#if	UNIX
		printf("Usage: arc {amufdxerplvtc}[biswnoq][g<password>]");
#endif
#if	_MTS
		printf("Parameters: {amufdxeplvtc}[biswnoq][g<password>]");
#endif
		printf(" <archive> [<filename> . . .]\n");
		printf("Where:	 a   = add files to archive\n");
		printf("	 m   = move files to archive\n");
		printf("	 u   = update files in archive\n");
		printf("	 f   = freshen files in archive\n");
		printf("	 d   = delete files from archive\n");
		printf("	 x,e = extract files from archive\n");
#if	!_MTS
		printf("	 r   = run files from archive\n");
#endif
		printf("	 p   = copy files from archive to");
		printf(" standard output\n");
		printf("	 l   = list files in archive\n");
		printf("	 v   = verbose listing of files in archive\n");
		printf("	 t   = test archive integrity\n");
		printf("	 c   = convert entry to new packing method\n");
		printf("	 b   = retain backup copy of archive\n");
#if	GEMDOS
		printf("	 h   = hold screen after finishing\n");
#endif
#if	_MTS
		printf("	 i   = suppress ASCII/EBCDIC translation\n");
#endif
#if	UNIX
		printf("	 i   = suppress image mode (translate EOL)\n");
#endif
		printf("	 s   = suppress compression (store only)\n");
		printf("	 w   = suppress warning messages\n");
		printf("	 n   = suppress notes and comments\n");
		printf("	 o   = overwrite existing files when");
		printf(" extracting\n");
		printf("	 q   = squash instead of crunching\n");
		printf("	 g   = Encrypt/decrypt archive entry\n");
		printf("\nAdapted from MSDOS by Howard Chu\n");
		/*
		 * printf("\nPlease refer to the program documentation for");
		 * printf(" complete instructions.\n"); 
		 */
#if	GEMDOS
		exitpause();
#endif
		return 1;
	}
	/* see where temp files go */
#if	!_MTS
	arctemp = calloc(1, STRLEN);
	if (!(arctemp2 = envfind("ARCTEMP")))
		arctemp2 = envfind("TMPDIR");
	if (arctemp2) {
		strncpy(arctemp, arctemp2, STRLEN - 16);
		arctemp[STRLEN - 17] = '\0';
		n = strlen(arctemp);
		if (arctemp[n - 1] != CUTOFF)
			arctemp[n] = CUTOFF;
	}
#if	UNIX
	else	strcpy(arctemp, "/tmp/");
#endif
#if	!MSDOS
	{
		strcat(arctemp, "AXXXXXX");
		int fd = mkstemp(arctemp);
		if (fd == -1)
		{
			fprintf(stderr, "Unable to create temporary files: %s\n", strerror(errno));
			exit(235);
		}
		close(fd);
		unlink(arctemp);
	}
#else
	strcat(arctemp, "$ARCTEMP");
#endif
#else
	guinfo("SHFSEP	", gotinf);
	sepchr[0] = gotinf[0];
	guinfo("SCRFCHAR", gotinf);
	tmpchr[0] = gotinf[0];
	arctemp = "-$$$";
	arctemp[0] = tmpchr[0];
#endif
	arctemp2 = NULL;

#if	!UNIX
	/* avoid any case problems with arguments */

	for (n = 1; n < num; n++)	/* for each argument */
		upper(arg[n]);	/* convert it to uppercase */
#else
	/* avoid case problems with command options */
	upper(arg[1]);		/* convert to uppercase */
#endif

	for (n = 1; n < num; n++)
	  if (strlen(arg[n]) > (STRLEN - 16))
	  {
	    fprintf(stderr, "Too long argument: %s\n", arg[n]);
	    exit(235);
	  }

	/* create archive names, supplying defaults */
#if	UNIX
	if (!stat(arg[2],&sbuf)) {
		if ((sbuf.st_mode & S_IFMT) == S_IFDIR)
			makefnam(arg[2],".arc",arcname);
		else
			strcpy(arcname,arg[2]);
	} else
		makefnam(arg[2],".arc",arcname);
#else
	makefnam(arg[2], ".ARC", arcname);
#endif
	/* makefnam(".$$$",arcname,newname); */
	sprintf(newname, "%s.arc", arctemp);
	makefnam(".BAK", arcname, bakname);

	/* now scan the command and see what we are to do */

	for (a = arg[1]; *a; a++) {	/* scan the option flags */
#if	!_MTS
		if (index("AMUFDXEPLVTCR", *a)) {	/* if a known command */
#else
		if (index("AMUFDXEPLVTC", *a)) {
#endif
			if (opt)/* do we have one yet? */
				arcdie("Cannot mix %c and %c", opt, *a);
			opt = *a;	/* else remember it */
		} else if (*a == 'B')	/* retain backup copy */
			keepbak = 1;

		else if (*a == 'W')	/* suppress warnings */
			warn = 0;
#if	!DOS
		else if (*a == 'I')	/* image mode, no ASCII/EBCDIC x-late */
			image = !image;
#endif
#if	GEMDOS
		else if (*a == 'H')	/* pause before exit */
			hold = 1;
#endif

		else if (*a == 'N')	/* suppress notes and comments */
			note = 0;

		else if (*a == 'O')	/* overwrite file on extract */
			overlay = 1;

		else if (*a == 'G') {	/* garble */
			password = a + 1;
			while (*a)
				a++;
			a--;
#if	_MTS
			etoa(password, strlen(password));
#endif
		} else if (*a == 'S')	/* storage kludge */
			nocomp = 1;

		else if (*a == 'K')	/* special kludge */
			kludge = 1;

		else if (*a == 'Q')	/* use squashing */
			dosquash = 1;

		else if (*a == '-' || *a == '/')	/* UNIX and PC-DOS
							 * option markers */
			;

		else
			arcdie("%c is an unknown command", *a);
	}

	if (!opt)
		arcdie("I have nothing to do!");

	/* get the files list set up */

	lnum = num - 3;		/* initial length of list */
	lst = (char **) calloc((lnum==0) ? 1:lnum,
				 sizeof(char *));	/* initial list */
	for (n = 3; n < num; n++)
		lst[n - 3] = arg[n];

	for (n = 0; n < lnum;) {/* expand indirect references */
		if (*lst[n] == '@')
			expandlst(n);
#if	GEMDOS		/* redirect stdout from the desktop...*/
		else if (*lst[n] == '>') {
			arctemp2 = (++lst[n]);
			lst[n] = lst[--lnum];	/* delete this entry */
			if (arctemp2[0] == '>') {
				append = 1;
				arctemp2++;
			}
			else	append = 0;
		}
#endif
		else
			n++;
	}
#if	GEMDOS
	if (arctemp2)
		freopen(arctemp2,append ? "a" : "w",stdout);
#endif

	/* Allocate space to I/O buffers */

	if (!(pinbuf = malloc(MYBUF)))
		arcdie("Not enough memory for input buffer.");
	if (!(outbuf = malloc(MYBUF)))
		arcdie("Not enough memory for output buffer.");
	if (!(pakbuf = malloc(2L*MYBUF)))
		arcdie("Not enough memory for packing buffer.");
	outend = outbuf + MYBUF - 1;
	
	
	/* act on whatever action command was given */

	switch (opt) {		/* action depends on command */
	case 'A':		/* Add */
	case 'M':		/* Move */
	case 'U':		/* Update */
	case 'F':		/* Freshen */
		addarc(lnum, lst, (opt == 'M'), (opt == 'U'), (opt == 'F'));
		break;

	case 'D':		/* Delete */
		delarc(lnum, lst);
		break;

	case 'E':		/* Extract */
	case 'X':		/* eXtract */
	case 'P':		/* Print */
		extarc(lnum, lst, (opt == 'P'));
		break;

	case 'V':		/* Verbose list */
		bose = 1;
	case 'L':		/* List */
		lstarc(lnum, lst);
		break;

	case 'T':		/* Test */
		tstarc();
		break;

	case 'C':		/* Convert */
		cvtarc(lnum, lst);
		break;
#if	!_MTS
	case 'R':		/* Run */
		runarc(lnum, lst);
		break;
#endif
	default:
		arcdie("I don't know how to do %c yet!", opt);
	}
#if	GEMDOS
	if (hold)
		exitpause();
#endif
	return nerrs;
}
static	VOID
expandlst(n)			/* expand an indirect reference */
	int		n;	/* number of entry to expand */
{
	FILE	       *lf, *fopen();	/* list file, opener */
	char		buf[100];	/* input buffer */
	int		x;	/* index */
	char	       *p = lst[n] + 1; /* filename pointer */

	if (*p) {		/* use name if one was given */
		makefnam(p, ".CMD", buf);
		if (!(lf = fopen(buf, "r")))
			arcdie("Cannot read list of files in %s", buf);
	} else
		lf = stdin;	/* else use standard input */

	for (x = n + 1; x < lnum; x++)	/* drop reference from the list */
		lst[x - 1] = lst[x];
	lnum--;

	while (fscanf(lf, "%99s", buf) > 0) {	/* read in the list */
		if (!(lst =(char **)realloc(lst, (lnum + 1) * sizeof(char *))))
			arcdie("too many file references");

		lst[lnum] = malloc(strlen(buf) + 1);
		strcpy(lst[lnum], buf); /* save the name */
		lnum++;
	}

	if (lf != stdin)	/* avoid closing standard input */
		fclose(lf);
}
